papers/libceed-paper/paper.rst

```diff
@@ -0,0 +1,72 @@
+
+In this work, we first introduce libCEED’s conceptual framework and C interface, and then illustrate its new Python interface, developed using the C Foreign Function Interface (CFFI) for Python. CFFI allows to reuse most of the C declarations and requires only a minimal adaptation of some of them. The C and Python APIs are mapped in a nearly 1:1 correspondence. For instance, data stored in the CeedVector structure are associated to arrays defined via the numpy package. In fact, since libCEED heavily relies on pointers and arrays to handle the data, a Python structure that resembles the C arrays is needed. In details, numpy arrays allow this correspondence obtained by passing the numpy array memory address as pointers to the libCEED C API.
+
+libCEED's API
+----------------------------------------------------------------------------------
+
+LibCEED's API provides the local action of the linear or nonlinear operator without assembling its sparse representation. Let us define the global operator as
+
+.. math::
+   :label: eq-operator-decomposition
+
+   A = P^T \underbrace{G^T B^T D B G}_{\text{libCEED}} P \, ,
+
+where :math:`P` is the parallel process decomposition operator (external to ``libCEED``) in which the degrees of freedom (DOFs) are scattered to and gathered from the different compute devices. The operator denoted by :math:`A_L = G^T B^T D B G` gives the local action on a compute node or process, where :math:`G` is a local element restriction operation that localizes DOFs based on the elements, :math:`B` defines the action of the basis functions (or their gradients) on the nodes, and :math:`D` is the user-defined pointwise function describing the physics of the problem at the quadrature points, also called the QFunction (see Fig. :ref:`fig-operator-decomp`).
+
+.. figure:: libCEED.png
+   :align: center
+   :figclass: bht
+
+   Operator decomposition.
+
+The mathematical formulation of QFunctions, described in weak form, is fully separated from the parallelization and meshing concerns. In fact, QFunctions, which can either be defined by the user or selected from a gallery of available built-in functions in the library, are pointwise functions that do not depend on element resolution, topology, or basis degree (selectable at run time). This easily allows :math:`hp`-refinement studies (where :math:`h` commonly denotes the average element size and :math:`p` the polynomial degree of the basis functions in 1D) and :math:`p`-multigrid solvers. libCEED also supports composition of different operators for multiphysics problems and mixed-element meshes (see Fig. :ref:`fig-schematic`). Currently, user-defined QFunctions are written in C and must be precompiled before running the Python script. The ultimate goal is for users to write only Python code. This will be achieved in the near future by using the Numba high-performance Python compiler.
```

Comment:
Perhaps
"Currently, user-defined QFunctions are written in C and must be precompiled as a foreign function library and loaded via ctypes."
and (if we show easy GPU support, which I think would be cool)
"However, these C QFunctions allow users to compute on GPU devices that libCEED supports."
(If OCCA is merged by then, that will be Nvidia + AMD)

Aside - Numba won't offer single source though, right? You can run Numba + CUDA, but it won't work with our CUDA backends.

Reply:
Attempted in [471a5cc](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/471a5cc68c77240ad12631faa9ccb5c17c6e93cd) (altough the diff is hard to read, even though I used a single line per paragraph, as discussed in the documentation PR)

Reply:
I'm not sure about your aside to be honest. Need to learn more

Reply:
Right now, the CUDA/MAGMA backends take the single source QFunctions in C source files and compile them as device kernels, with appropriate change to the CEED_QFUNCTION macro. If I understand the capabilities correctly, we can use Numba to make a function pointer for the CPU backends but that code won't be compatible with single source QFunctions. We might be able to make a backend that works with numba CUDA, though I think it would take more work than on the CPU side as having the QFunction source is more integral to the CUDA backends.

Reply:
I see...

Reply:
We should refer to JAX instead of (or in addition to) Numba. Notes here say sharing of device arrays is supported. https://jax.readthedocs.io/en/latest/CHANGELOG.html?highlight=__cuda_array_interface__#jax-0-1-59-february-11-2020

I believe the cuda-shared and magma backends would be able to use JAX or Numba qfunctions, but cuda-gen would not because the *.cu source is not available (so it would at least need significant JIT shenanigans, if supportable at all).

Reply:
Mentioned both possibilities in [c5a9041](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/c5a9041c8f26486f687aaad0b68c75bde90ed197)

Reply:
/gpu/cuda/ref /gpu/cuda/reg and /gpu/cuda/shared would still require some modification to replace the compilation of the QF with getting the kernel from JAX/Numba but it looks easier. I agree that I'm not sure if /cuda/gen is possible.